### DISCUSSION TITLE
Allow trimming and inlining critical CSS

### DISCUSSION THREAD

**Discussion by Demivan** [vikejs/vike#1975] [Discussion] [answered]
### Description

Right now, there is no easy way to integrate https://github.com/danielroe/beasties to remove not used CSS from the page when pre-rendering.

**Comment by brillout** [comment]
Feel free to elaborate on what's missing on Vike's side for you to integrate Beasties.

**Reply by Demivan** [reply]
Vike inserts CSS links internally, and there is no hook to modify HTML after it is rendered. Beasties needs the final HTML with CSS links (or inlined CSS) to be able to inline CSS, trim not used classes, and modify CSS imports.

Currently, I'm doing this in `onRenderHtml()`: 

```ts
	const assets = await pageContext.__getPageAssets()
	const css = assets
		.filter(asset => asset.assetType === 'style' && asset.isEntry)
		.map(asset => asset.src)
		.map(css => `<link rel="stylesheet" href="${css}">`)
		.join('\n')

	const beasties = new Beasties({
		path: path.join(process.cwd(), 'dist/client'),
	})

	const cleanPageHtml = await beasties.process(`<head>${css}</head><body>${pageHtml}</body>`)

	const headHtml = cleanPageHtml.match(/<head>([\s\S]*)<\/head>/)[1]!
	const bodyHtml = cleanPageHtml.match(/<body>([\s\S]*)<\/body>/)[1]!

	const documentHtml = escapeInject`<!DOCTYPE html>
<head>...${dangerouslySkipEscape(headHtml)}...</head>
<body>...${dangerouslySkipEscape(bodyHtml)}...</body>
</html>`

	const injectFilter = (assets: InjectFilterEntry[]) => {
		assets.forEach((asset) => {
			// Do not inject styles
			if (asset.assetType === 'style') {
				asset.inject = false
			}
		})
	}
```

This causes `We recommend against not injecting..` warnings and uses internal Vike `__getPageAssets()` api. And overall seems like a hacky solution.

Vike needs a hook for final processing of the generated HTML. Or even native Beasties integration. I think it would be nice to be able to exclude not used CSS when prerendering.

**Reply by brillout** [reply, marked as answer]
I agree: https://github.com/vikejs/vike/issues/1979. Creating a `vike-beasties` should be easy then.

Another less hacky workaround would be to crawl HTML files in `dist/client/` and run Beasties against each HTML file.
